notebook/tests/launchnotebook.py

```diff
@@ -90,23 +91,41 @@ def request(cls, verb, path, **kwargs):
             url_path_join(cls.base_url(), path),
             **kwargs)
         return response
-
+    
     @classmethod
     def setup_class(cls):
-        cls.home_dir = TemporaryDirectory()
-        data_dir = TemporaryDirectory()
+        cls.tmp_dir = TemporaryDirectory()
+        def tmp(*parts):
+            path = os.path.join(cls.tmp_dir.name, *parts)
+            try:
+                os.makedirs(path)
+            except OSError as e:
+                if e.errno != errno.EEXIST:
+                    raise
+            return path
+        
+        cls.home_dir = tmp('home')
+        data_dir = cls.data_dir = tmp('data')
+        config_dir = cls.config_dir = tmp('config')
+        runtime_dir = cls.runtime_dir = tmp('runtime')
         cls.env_patch = patch.dict('os.environ', {
-            'HOME': cls.home_dir.name,
+            'HOME': cls.home_dir,
             'PYTHONPATH': os.pathsep.join(sys.path),
-            'IPYTHONDIR': pjoin(cls.home_dir.name, '.ipython'),
-            'JUPYTER_DATA_DIR' : data_dir.name
+            'IPYTHONDIR': pjoin(cls.home_dir, '.ipython'),
+            'JUPYTER_NO_CONFIG': '1', # needed in the future
+            'JUPYTER_CONFIG_DIR' : config_dir,
+            'JUPYTER_DATA_DIR' : data_dir,
+            'JUPYTER_RUNTIME_DIR': runtime_dir,
         })
         cls.env_patch.start()
-        cls.path_patch = patch.object(jupyter_core.paths, 'SYSTEM_JUPYTER_PATH', [])
+        cls.path_patch = patch.multiple(
+            jupyter_core.paths,
+            SYSTEM_JUPYTER_PATH=[tmp('share', 'jupyter')],
+            ENV_JUPYTER_PATH=[tmp('env', 'share', 'jupyter')],
+            SYSTEM_CONFIG_PATH=[tmp('etc', 'jupyter')],
+            ENV_CONFIG_PATH=[tmp('env', 'etc', 'jupyter')],
+        )
         cls.path_patch.start()
-        cls.config_dir = TemporaryDirectory()
-        cls.data_dir = data_dir
```

Comment:
There are some tests expecting `self.data_dir` to be a TemporaryDirectory instance, which are now failing.